### PR TITLE
Call a global #on_failure hook when the flow fails

### PR DESCRIPTION
**Pull request by waiting-for-dev**
Most of the time, individual operations are responsible for handling their own failures. However, there are cases when it makes sense to handle failures globally, for example, when you want to log the error for a given flow.

When using the raw behavior, i.e., when we don't prepend around methods, that can be easily handled manually:

```ruby
class CreateUser < Dry::Operation
  skip_prepending

  def call(input)
    steps do
      attrs = step validate(input)
      step persist(attrs)
      user
    end.tap do |result|
      log_failure(result.failure) if result.failure?
    end
  end

  # ...
end
```

However, by automatically wrapping around `#steps` we gain focus on the happy path, but we lose the ability to handle global failures.

Because of that, we introduce an `#on_failure` hook that is only called when using the prepended behavior on a failing flow. The method accepts the unwrapped failure extracted from the result object.

```ruby
class CreateUser < Dry::Operation
  def call(input)
    attrs = step validate(input)
    step persist(attrs)
    user
  end

  private

  def on_failure(failure)
    log_failure(failure)
  end

  # ...
end
```

`#on_failure` can also take a second optional argument, which will be assigned to the prepended method's name. That's useful when we're defining more than one flow in a single class.

```ruby
class UserFlows < Dry::Operation
  operate_on :create_user, :delete_user

  # ...

  private

  def on_failure(failure, flow_name)
    case flow_name
    when :create_user
      # ...
    when :delete_user
      # ...
    end
  end

  # ...
end
```

The calling of the hook is done via an injected result handler lambda. At some point, we might want to make this behavior configurable.

### CODE REVIEW

**Comment by waiting-for-dev** [comment]
@timriley, the `StepsMethodPrepender` module is now doing more than just prepending with `#steps`, as it's also invoking the result handler. Do we want to rename it? What do you think about something along the lines of `Streamliner`, `Enhancer` or `Facilitator`?

File: lib/dry/operation/class_context/steps_method_prepender.rb
Comment (timriley) [inline lib/dry/operation/class_context/steps_method_prepender.rb:13]:
```suggestion
          return if result.success? || !(instance.methods.include?(FAILURE_HOOK_METHOD_NAME) || instance.private_methods.include?(FAILURE_HOOK_METHOD_NAME))
```
```suggestion
          return if result.success? || !instance.methods.include?(FAILURE_HOOK_METHOD_NAME)
```

Would it be worth doing this to allow `on_failure` to be private? If I'm thinking about the external interface to operation objects, certainly `#on_failure` doesn't feel like something we want to make callable externally.

Comment (timriley) [reply lib/dry/operation/class_context/steps_method_prepender.rb:13]: I also wonder if we could cache the result of this `methods`/`private_methods` lookup eventually...

Actually, this makes me wonder, instead of _looking up a possible on_failure method_, could we instead _implement_ a default empty `on_failure` method ourselves? We could define it on the instance whenever `StepsMethodPrepender` is included.

This would allow this result handler to be much more confident. It would mean this first line becomes a simpler `return if result.success?`.

File: lib/dry/operation/class_context/steps_method_prepender.rb
Comment (timriley) [inline lib/dry/operation/class_context/steps_method_prepender.rb:46]: What I would love here is we can somehow _name the individual steps themselves_ and have it passed to the result handler. That way it could do even finer-grained handling of errors.

Perhaps we could consider an optional syntax like `step :step_name, value`. Converting our example from our inline docs:

```ruby
def call(input)
  attrs = step :validate, validate(input)
  user = step :persist, persist(attrs)
  step :notify, notify(user)
  user
end
```

Definitely not for this PR, but I'm keen for your thoughts! Now that we have this result handling, it feels like there is more potential value in (optionally) naming steps like this.

**Review by timriley — COMMENTED** [review]
This is looking really good! I've left a couple of notes that I think we would be good to address before we merge this (and a few thoughts that are for well beyond this PR, but they felt relevant to include along the way) — let me know what you think!

File: lib/dry/operation.rb
Comment (waiting-for-dev) [reply lib/dry/operation.rb]: Yeah, I'm a bit torn on this as we use `value` to reference what is inside a `Success` (i.e., `value!` returns error when called on a `Failure`). I changed it, but I'd like to find a better name, although it's true that `failure` is neither good as we're passing the wrapped *thing*.

File: lib/dry/operation/errors.rb
Comment (waiting-for-dev) [reply lib/dry/operation/errors.rb]: Sure thing. I'll follow-up in another PR.

File: lib/dry/operation/class_context/steps_method_prepender.rb
Comment (waiting-for-dev) [reply lib/dry/operation/class_context/steps_method_prepender.rb]: Makes sense. Changed!

File: lib/dry/operation/class_context/steps_method_prepender.rb
Comment (waiting-for-dev) [reply lib/dry/operation/class_context/steps_method_prepender.rb]: I changed it to use `#respond_to?` to make sure it works for any kind of method (public, private, inherited). What do you think? I also added a test to make sure it works with `private` methods.

> Actually, this makes me wonder, instead of looking up a possible on_failure method, could we instead implement a default empty on_failure method ourselves? We could define it on the instance whenever StepsMethodPrepender is included.

I'd like to have the `RESULT_HANDLER` injectable as it is now, this way we make all the magic in the gem configurable :/ WDYT?

File: lib/dry/operation/class_context/steps_method_prepender.rb
Comment (waiting-for-dev) [reply lib/dry/operation/class_context/steps_method_prepender.rb]: I'm also torn here. Not sure we should make this as something built-in in the library as you can already use Ruby + pattern matching for that, so maybe just providing examples would be enough. I mean, we could have the `validate` method return a `[:validation_error, errors]` tuple. I realize that's not too idiomatic, though.

File: lib/dry/operation.rb
Comment (timriley) [reply lib/dry/operation.rb]: `failure_value`?

File: lib/dry/operation/class_context/steps_method_prepender.rb
Comment (timriley) [reply lib/dry/operation/class_context/steps_method_prepender.rb]: Thanks for incorporating private methods :)

I wasn't suggesting we change the injection of `RESULT_HANDLER`, instead what I was suggesting was this:

```
module Dry
  class Operation
    private

    def on_failure
      # empty method, always there by default
    end
  end
end
```

And inside `RESULT_HANDLER`, we no longer need to check for the presence of `#on_failure`, because we can rely on it always being there courtesy of the above. So this:

```ruby
        RESULT_HANDLER = lambda do |instance, method, result|
          return if result.success? ||
                    !(instance.methods + instance.private_methods).include?(
                      FAILURE_HOOK_METHOD_NAME
                    )
```

Becomes this:

```ruby
        RESULT_HANDLER = lambda do |instance, method, result|
          return if result.success?
```

Does that make sense? Create the method once, inside the gem code, and we no longer need to do introspection to find it.

I made those changes locally and all tests still passed.

---

While I'm here, I do have one question regarding `RESULT_HANDLER` being injectable. Who do you expect to be injecting alternatives? End users? Given of how fundamental that handler is to the behaviour of the library, do we actually _want_ it to ever be replaced? (In addition, `StepsMethodPrepender` is multiple layers removed from the classes that users are interacting with directly, anyway).

File: lib/dry/operation/class_context/steps_method_prepender.rb
Comment (timriley) [reply lib/dry/operation/class_context/steps_method_prepender.rb]: Yeah. Let's see about this in the future sometime. We can have a play and see how it feels.

**Review by timriley — APPROVED** [review]
Left a couple extra notes and [one suggestion for a change](https://github.com/dry-rb/dry-operation/pull/14#discussion_r1648771072) — once you've considered that, then I think this is good to merge! Thanks @waiting-for-dev!

File: lib/dry/operation.rb
Comment (waiting-for-dev) [reply lib/dry/operation.rb]: Good enough 🙂 Changed!

File: lib/dry/operation/class_context/steps_method_prepender.rb
Comment (waiting-for-dev) [reply lib/dry/operation/class_context/steps_method_prepender.rb]: > I wasn't suggesting we change the injection of RESULT_HANDLER, instead what I was suggesting was this:

However, in the case that another result handler was in use, it would make no sense that we're defining the `#on_failure` in the class.

> While I'm here, I do have one question regarding RESULT_HANDLER being injectable. Who do you expect to be injecting alternatives? End users? Given of how fundamental that handler is to the behaviour of the library, do we actually want it to ever be replaced? (In addition, StepsMethodPrepender is multiple layers removed from the classes that users are interacting with directly, anyway).

You're right, but `StepsMethodPrepender` can be completely disabled with `.skip_prepending`. I think it's good to avoid having any magic as something you can't, at least, opt-out. It's not only respectful with the user, but it also guides us through coming up with a better architecture. I think the `#on_failure` hook is the same case here. Let me know what you think and let's take a final decision 🙂

File: lib/dry/operation/errors.rb
Comment (waiting-for-dev) [reply lib/dry/operation/errors.rb]: Done here by @cllns 🙂

File: lib/dry/operation/class_context/steps_method_prepender.rb
Comment (timriley) [reply lib/dry/operation/class_context/steps_method_prepender.rb]: @waiting-for-dev Thanks for explaining that, you've convinced me, happy to leave this as-is! 🚀

**Comment by timriley** [comment]
@waiting-for-dev Have [answered](https://github.com/dry-rb/dry-operation/pull/14#discussion_r1649625055) the final outstanding question, this looks good to merge!

File: lib/dry/operation/class_context/steps_method_prepender.rb
Comment (waiting-for-dev) [reply lib/dry/operation/class_context/steps_method_prepender.rb]: @timriley, thanks for keeping an open mind here, your feedback is always so valuable! ❤️

**Pull request merged**
